kura/examples/org.eclipse.kura.demo.modbus/src/main/java/org/eclipse/kura/demo/modbus/ModbusExample.java

```diff
@@ -44,99 +86,221 @@ public void unsetModbusProtocolDeviceService(ModbusProtocolDeviceService modbusS
         s_logger.info("Modbus polling thread killed");
 
         if (this.m_protocolDevice != null) {
             try {
                 this.m_protocolDevice.disconnect();
             } catch (ModbusProtocolException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
+                s_logger.error("Failed to disconnect : {}", e.getMessage());
             }
         }
         this.configured = false;
+
+        // Releasing the CloudApplicationClient
+        s_logger.info("Releasing CloudApplicationClient for {}...", APP_ID);
+        this.m_cloudClient.release();
+
     }
 
     public void updated(Map<String, Object> properties) {
         s_logger.info("updated...");
-        this.m_properties = properties;
-        modbusProperties = getModbusProperties();
-        this.m_slaveAddr = Integer.valueOf(modbusProperties.getProperty("slaveAddr")).intValue();
         this.configured = false;
+        doUpdate(properties);
     }
 
-    private int bcd2Dec(int bcdVal) {
-        byte bcd = (byte) bcdVal;
-        int decimal = (bcd & 0x000F) + ((bcd & 0x000F0) >> 4) * 10 + ((bcd & 0x00F00) >> 8) * 100
-                + ((bcd & 0x0F000) >> 12) * 1000 + ((bcd & 0xF0000) >> 16) * 10000;
+    // ----------------------------------------------------------------
+    //
+    // Private Methods
+    //
+    // ----------------------------------------------------------------
 
-        return decimal;
-    }
+    /**
+     * Called after a new set of properties has been configured on the service
+     */
+    private void doUpdate(Map<String, Object> properties) {
+        try {
+
+            for (String s : properties.keySet()) {
+                s_logger.info("Update - " + s + ": " + properties.get(s));
+            }
+
+            // cancel a current worker handle if one if active
+            if (this.m_handle != null) {
+                this.m_handle.cancel(true);
+            }
+
+            if (this.m_protocolDevice != null) {
+                try {
+                    this.m_protocolDevice.disconnect();
+                } catch (ModbusProtocolException e) {
+                    s_logger.error("Failed to disconnect : {}", e.getMessage());
+                }
+            }
+            this.configured = false;
+
+            this.m_properties = properties;
+            modbusProperties = getModbusProperties();
+            if (this.m_properties.get(PUBLISH_TOPIC_PROP_NAME) != null) {
+                this.m_topic = (String) this.m_properties.get(PUBLISH_TOPIC_PROP_NAME);
+            }
+            if (this.m_properties.get(PUBLISH_INTERVAL) != null) {
+                this.m_publish_interval = (Integer) this.m_properties.get(PUBLISH_INTERVAL);
+            }
+
+            if (this.m_properties.get(INPUT_ADDRESS) != null) {
+                inputaddr = (Integer) this.m_properties.get(INPUT_ADDRESS);
+            }
+            else inputaddr=0;
```

Comment:
You don't need the else part as `inputaddr` is already initialised to 0. If there doesn't exist any property for `INPUT_ADDRESS`, it will always be 0.